internal/installercache/installercache.go

```diff
@@ -0,0 +1,91 @@
+package installercache
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type installers struct {
+	sync.Mutex
+	releases map[string]*release
+}
+
+type release struct {
+	sync.Mutex
+	path string
+}
+
+var cache installers = installers{
+	releases: make(map[string]*release),
+}
+
+// Get returns a release resource for the given release ID
+func (i *installers) Get(releaseID string) *release {
+	i.Lock()
+	defer i.Unlock()
+
+	r, present := i.releases[releaseID]
+	if !present {
+		r = &release{}
+		i.releases[releaseID] = r
+	}
+	return r
+}
+
+// Get returns the path to an openshift-baremetal-install binary extracted from
+// the referenced release image. It is safe for concurrent use. A cache of
+// binaries is maintained to reduce re-downloading of the same release.
+func Get(releaseID, cacheDir, pullSecret string, log logrus.FieldLogger) (string, error) {
+	r := cache.Get(releaseID)
+	r.Lock()
+	defer r.Unlock()
+
+	// cache miss
+	if r.path == "" {
+		// write pull secret to a temp file
+		ps, err := ioutil.TempFile("", "registry-config")
+		if err != nil {
+			return "", err
+		}
+		defer func() {
+			ps.Close()
+			os.Remove(ps.Name())
+		}()
+		_, err = ps.Write([]byte(pullSecret))
+		if err != nil {
+			return "", err
+		}
+		// flush the buffer to ensure the file can be read
+		ps.Close()
+
+		workdir := filepath.Join(cacheDir, releaseID)
+		log.Infof("extracting openshift-baremetal-install binary to %s", workdir)
+		err = os.MkdirAll(workdir, 0755)
+		if err != nil {
+			return "", err
+		}
+
+		// extract openshift-baremetal-install binary from the release
+		cmd := exec.Command("oc", "adm", "release", "extract", "--command=openshift-baremetal-install",
```

Comment:
As much as I'm enjoying this cache implementation, is this really better than building all of the releases we support into the image?

I probably missed this conversation the first time it happened, but I have a few concerns:
- The main trade-off we're making here is faster build/deploy for slower ignition generation when we have a cache miss. I'm not convinced this is a good trade.
- We're using the customer's pull secret here to download the installer, right? I'm not sure if that could cause a problem but it seems ... strange. Maybe we should have some kind of system-level credential for doing this.
- "No code is more stable than no code". I'm confident in @mhrivnak 's implementation, but just pulling the binaries into the image is certain to be easier to maintain.

Reply:
I agree with all of your points, and I'm a big fan of less code. The simplest and most robust option is to say that a given release of assisted-service supports a set of OCP releases, and bake in their installer binaries. Here are the downsides:

* Lifecycle-wise, there's no fundamental reason a given release of assisted-service should be limited to a set of OCP releases. Building and releasing a new assisted-service version just to add or subtract an OCP release is likely to be an unnecessary amount of overhead. The set of OCP releases it works with is a better conceptual fit as configuration.
* For on-prem use, we can imagine letting a customer mirror any 4.Y release they want into their local registry, and have assisted-service work with that. We don't want to make them pick a release up-front and have to later download the entire thing over again just to install a different release.
* The image gets large. Particularly for developers without amazing bandwidth at home, moving images around with these giant (300MB+ each) binaries included slows them down.

The implementation here still allows the person deploying assisted-service to either build an additional layer on the container image that adds their desired binaries, or mount in some other storage at runtime that contains one or more binaries.

In terms of slowing down ignition generation when there's a cache miss, hopefully in production, assisted-service will be able to download the binary quickly. If there's an operational reason why that's not possible, it would be good to discuss. And of course there's still the opportunity in production to pre-warm the cache as described above.

For the pull secret, I'm just not aware of a better option for the moment. I agree it's conceptually not the best.

Reply:
> The implementation here still allows the person deploying assisted-service to either build an additional layer on the container image that adds their desired binaries, or mount in some other storage at runtime that contains one or more binaries.

This is a great point. We should monitor for outliers in generation time after this is merged and see if something like this is necessary specifically for the assisted-service deployed in cloud.redhat.com. We could then also maintain which releases are available separately from the running image.

> For the pull secret, I'm just not aware of a better option for the moment. I agree it's conceptually not the best.

Yeah, I definitely don't know what else is available to us. @rollandf @ybettan do you know if there is any other auth we could use for this operation?

Reply:
Also I really only posted this to start a conversation. I fully expect this to get merged with the download/caching solution in place.

Reply:
I also have a bit of documentation to push that I've been pecking away at today between test runs and builds, explaining how and why to add storage for these purposes.

Reply:
Pull secrets have given all-or-nothing access since they started being used. Are there openshift releases we're now restricting by individual pull secrets?

Reply:
It's also worth noting that a user can't download the binary from this cache. All they can get access to is ignition files generated by a particular installer.